src/test/run-pass/arbitrary_enum_discriminant.rs

```diff
@@ -0,0 +1,55 @@
+#![feature(arbitrary_enum_discriminant, const_raw_ptr_deref, test)]
+
+extern crate test;
+
+use test::black_box;
+
+#[allow(dead_code)]
+#[repr(u8)]
+enum Enum {
+    Unit = 3,
+    Tuple(u16) = 2,
+    Struct {
+        a: u8,
+        b: u16,
+    } = 1,
+}
+
+impl Enum {
+    const unsafe fn tag(&self) -> u8 {
+        *(self as *const Self as *const u8)
+    }
+}
+
+#[allow(dead_code)]
+#[repr(u8)]
+enum FieldlessEnum {
+    Unit = 3,
+    Tuple() = 2,
+    Struct {} = 1,
+}
+
+fn main() {
+    const UNIT: Enum = Enum::Unit;
+    const TUPLE: Enum = Enum::Tuple(5);
+    const STRUCT: Enum = Enum::Struct{a: 7, b: 11};
+
+    // Ensure discriminants are correct during runtime execution
+    assert_eq!(3, unsafe { black_box(UNIT).tag() });
```

Comment:
Why `black_box`? That has no bearing on whether this is CTFE or not.

Reply:
it does, because the code may get promoted otherwise. Due to the method call (`tag`) that doesn't happen though, but the `UNIT` may still get promoted.

Reply:
The `UNIT` is a const, it doesn't need promotion.

Reply:
Yea, there's no notable difference here

Reply:
Since `UNIT` is const, and `Unit::tag` is a const fn, I worried that rustc (or a future rustc) could CTFE `UNIT.tag()`. Wrapping `UNIT` with `black_box` introduces a boundary between the value and the method call, hopefully preventing this (since this particular test is to ensure that discriminants are correct during _runtime_ execution).

Reply:
Note that `black_box` is purely an optimization hint and not a semantic guarantee -- operationally you can only assume that it is the identity function.

Reply:
Sure, but all we need here is that it is definitely not getting promoted statically.

Reply:
I'm confused, has the conversation here been resolved, or not? @RalfJung, do you still think the call to `black_box` should be removed?

Reply:
I am fine with keeping it.